Transcoding via FFmpeg API

j0sh opened the pull request:
Commits roughly broken down by feature for easier review.

j0sh commented:
Ready for review.

ericxtang commented on ffmpeg/ffmpeg.go:
```diff
@@ -27,6 +33,37 @@ func RTMPToHLS(localRTMPUrl string, outM3U8 string, tmpl string, seglen_secs str
+func Transcode(input string, ps []VideoProfile) error {
+	inp := C.CString(input)
+	params := make([]C.output_params, len(ps))
+	for i, param := range ps {
```
I think we have to put the transcode call in a go routine somewhere in this call stack.  Otherwise it'll block until the work finishes.  Maybe the best place is in `Livepeernode.TranscodeAndBroadcast()` - just pointing it out so we don't forget.

j0sh replied:
Sure -- I believe the old behavior was synchronous as well. If the blocking has been a problem with with go-livepeer all along, then we can change it there.

ericxtang commented on ffmpeg/lpms_ffmpeg.c:
```diff
@@ -102,3 +140,588 @@ int lpms_rtmp2hls(char *listen, char *outf, char *ts_tmpl, char* seg_time)
+    vc = avcodec_alloc_context3(codec);
+    if (!vc) em_err("Unable to alloc video encoder\n"); // XXX shld be optional
+    octx->vc = vc;
+    vc->width = av_buffersink_get_w(octx->vf.sink_ctx);
```
Anywhere I can learn about the buffer source / buffer sink API?

j0sh replied:
Probably the canonical source for those specific functions is the header documentation, the comments are fairly thorough: https://github.com/FFmpeg/FFmpeg/blob/master/libavfilter/buffersink.h

ericxtang commented on ffmpeg/lpms_ffmpeg.c:
```diff
@@ -102,3 +140,588 @@ int lpms_rtmp2hls(char *listen, char *outf, char *ts_tmpl, char* seg_time)
+        encoder = octx->vc;
+        filter = &octx->vf;
+      } else if (ist->index == ictx.ai && ictx.ai) {
+        ost = octx->oc->streams[!!ictx.vc];
```
Why !!ictx.vc here instead of something like `octx->oc->streams[1]`?

j0sh replied:
Accommodates the case where we don't have video. Video streams are always at index 0, but audio may be 0 or 1 depending on whether `ictx.vc` is non-null.

j0sh replied:
Added a comment explaining as much, and test cases to ensure that it actually handles no-audio and no-video correctly. Note that this line contains a bug:

`} else if (ist->index == ictx.ai && ictx.ai) {`

which should be `} else if (ist->index == ictx.ai && ictx.ac) {`

ericxtang commented on ffmpeg/lpms_ffmpeg.c:
```diff
@@ -102,3 +140,588 @@ int lpms_rtmp2hls(char *listen, char *outf, char *ts_tmpl, char* seg_time)
+#undef proc_err
+}
+
+#define MAX_OUTPUT_SIZE 3
```
Why use MAX_OUTPUT_SIZE if we already have nb_outputs?

j0sh replied:
Good catch -- we preallocate MAX_OUTPUT_SIZE output contexts (on the stack), but we're missing the check for nb_outputs <= MAX_OUTPUT_SIZE.

j0sh commented:
Happy to explain whatever else about the code that may need clarification.

Agreed that the filtering API is a bit unexplored by the community, partially due to its relative new-ness. It's also a bit cumbersome to use compared to straight swscale and swresample, but I think the flexibility offered by filters will pay off in the long run. Filters have already been useful to avoid re-implementing fps handling, as well as certain adjustments on the audio (but the extra audio filters are disabled in the current code, aside from resampling).

For me, the best source of information has been the lavfi header documentation itself [1] , plus the docs on specific filters themselves [2] and the sample code [3][4].

[1] https://github.com/FFmpeg/FFmpeg/blob/master/libavfilter/avfilter.h
[2] http://ffmpeg.org/ffmpeg-filters.html
[3] https://github.com/FFmpeg/FFmpeg/blob/master/doc/examples/filtering_video.c
[4] https://github.com/FFmpeg/FFmpeg/blob/master/doc/examples/filter_audio.c

j0sh commented:
Fixed / clarified a couple of the points that were mentioned, and added tests for those cases to boot. Which is a good thing, because the tests uncovered bugs (which have also been fixed). There is also a small memory leak sometimes (about one frame's worth of memory) but will track that down.

ericxtang commented:
🚢  after addressing a small issue - **I ran test.sh in devenv and kept getting a left-over ts segment in `/transcoder`.  Seems like it's not being properly cleaned up**

This has been a very educational code review process for me.  I still can't say I'd be able to build this from scratch, but at least now I understand some of the basic concepts in libav.

This note was helpful for me as I was going through the codebase - 

open_output

* av_guess_format
* avformat_alloc_output_context2
* If we have input video context
    * get the x264 codec (which is also the encoder)
    * vc = avcodec_alloc_context3 based on the codec
    * set vc's parameters based on octx (width, height, framerate, time_base, thread count, pix_fmt)
        * some of these are retrieved by calling av_buffersink_get_x.  
    * avcodec_open2 (what is this for? why do you have to “open” the codec?)
    * avformat_new_stream(call on the output context)
    * avcodec_parameters_from_context
* If we have input audio context
    * get the acc codec (which is also the encoder)
    * ac = avcodec_alloc_context3 - open the codec
    * set ac's parameters based on octx
        * a lot of these are retrieved by calling av_buffersink_get_x.
    * avcodec_open2
    * avformat_new_stream
    * avcodec_parameters_from_context
    * set timebase
* avio_open
* av_format_write_header

open_input (initialize everything in input_ctx)

* avformat_open_input - create ic (input AVFormatContext)
* avformat_find_stream_info
* open video decoder
    * av_find_best_stream (get the video stream)
    * vc = avcodec_alloc_context3 (AVCodecContext)
    * avcodec_parameters_to_context(vc, ic→streams[ctx→vi]→codecpar
    * av_opt_set_int (what does this do?)
    * avcodec_open2
* open audio decoder
    * av_find_best_stream (get the audio stream)
    * ac = avcodec_alloc_context3
    * avcodec_parameters_to_context
    * avcodec_open2

init_video_filters

* allocates a filter graph (avfilter_graph_alloc)
* create the “source” of the filter by adding a buffer to the graph, passing string parameters to it
* create the “sink” of the filter by adding a buffer sink to the graph, where we will get our filtered frames from
* use the magical function call “avfilter_graph_parse_ptr” to create the graph with the string-based filter graph, a filter input and a filter output
* use avfilter_graph_config to make sure the graph is valid
* allocate the filter output frame with av_frame_alloc

init_audio_filters

* Pretty much works the same way as init_video_filters

process_in (put everything into the frame?)

* av_init_packet
* loop
    * av_read_frame(ictx→ic, pkt) - read from input context into the packet
    * pkt has a stream_index → it can be either a video packet or an audio packet
    * get the decoder depending on the pkt type (either video or audio decoder)
    * avcodec_send_packet(put the packet into the decoder)
    * avcodec_receive_frame(put the decoder content into the frame)
* at the end (AVERROR_EOF), flush the decoder
    * call avcodec_send_packet with NULL
    * call avcodec_receive_frame
    * why set pkt→stream_index?
    * why not call dec_cleanup after the flush?

process_out

* depending on if the filter is active, we do “filtering” on the input frame
    * av_frame_copy_props(frame, inf) → copy the inf into frame
    * av_buffersrc_add_frame_flags (what does this do?)
    * av_buffersink_get_frame(get the frame out of the sink_ctx?)
    * tb = av_buffersink_get_frame_time_base(filter→sink_ctx)
* av_init_packet
* if the encoder is set (when should we NOT set the encoder?)
    * avcodec_send_frame
    * avcodec_receive_packet
* set pkt.pts, pkt.dts, pkt.duration
* av_interleaved_write_frame (from packet into octx->oc)

lpms_transcode

* open_input with ictx (initialize every thing in ictx)
* memset outputs
* for each output profile
    * sets context params (fname, width, height, bitrate, fps)
    * init_video_filters
    * init_audio_filters
    * open_output (initialize everything in octx)
* av_init_packet
* av_frame_alloc
* while loop
    * process_in → demux, decode (put everything into the frame)
    * get input stream (could be video or audio)
    * for each output profile
        * get output stream, encoder, and filter (could be video or audio, depending on input stream)
        * process_out
* flush
    * process_out with NULL (instead of frame)
    * av_interleaved_write_frame on oc
    * av_write_trialer on oc

j0sh commented:
Thanks for the review and the questions, helps to re-visit some of the assumptions I may have been making when writing this.

> I ran test.sh in devenv and kept getting a left-over ts segment in /transcoder. Seems like it's not being properly cleaned up

Good catch, this was actually a latent issue in the old code as well -- if ffmpeg failed to start for any reason then it would have failed to clean up. Fixed.

@ericxtang  what do you think about changing this further to stop creating temporary files in `ffmpeg_segment_transcoder` and just take an existing filename? This would require minor changes to go-livepeer as well.

> avcodec_open2 (what is this for? why do you have to “open” the codec?)

There are two parts here: the AVCodec and the AVCodecContext.

* The AVCodec identifies the specific codec and its constraints (codec name, media type, parameters that the en/decoder supports such as the selection of pixel formats or sample rates, etc).

* The AVCodecContext is the actual instantiation of the de/encoder that is operated on. So we first find the AVCodec, then instantiate the AVCodecContext based on the parameters we need (and what the codec supports).

> av_opt_set_int (what does this do?)

This sets the decoded frames to be reference counted rather than following the old semantics of copying everything if some data needs to be retained outside the decode process. Actually it's deprecated right now due to our usage of the receive_frame API; it was cargo-culted from some libavfilter examples. Removed.

We actually should be setting all our optional fields (bitrate, sample rate, etc etc) via AVOption, added some more info here https://github.com/livepeer/lpms/issues/55#issuecomment-369030155

> init_audio_filters Pretty much works the same way as init_video_filters

Eventually, we should should look into refactoring this so there's less boilerplate (avfilter_graph_parse2 looks promising) and more shared code between the audio and video filters.

> why set pkt→stream_index?

When we flush, it's because we've EOF'd on the input. However, we may still have some frames pending, so setting pkt->stream_index during flushing is how we identify which (input) stream the flushed frame belongs to, so we can select the correct encoders for the output after the process_in function returns.

Normally doing it this way is safer than poking inside AVFrame itself to determine the content type (whether audio/video), because decoded AVFrames don't carry any information about the source packet/stream. This is problematic for eg, multiple audio or video streams in the same output container.

> why not call dec_cleanup after the flush?

dec_cleanup releases the reference to the AVPacket that was read during input, but when flushing we don't have any more AVPackets to read. Rather, we're trying to extract the last remaining frames that may be buffered within the decoder.

> av_buffersrc_add_frame_flags (what does this do?)

This is a rather confusingly named function, but it should be read like av_buffersrc_add_frame_{with}_flags . All it does is feed the buffersrc with the frame and some additional flags. Changed it to use a helper function that's a bit less explicit, but follows the same semantics: `av_buffersrc_write_frame` .

> av_buffersink_get_frame(get the frame out of the sink_ctx?)

Yep, gets the frame out of the filter graph (the buffersink being the 'exit point' for that).

> if the encoder is set (when should we NOT set the encoder?)

Right now it should always be set unless the file doesn't actually have an audio or video stream. That check might have been a bit of leftover debugging.

In general we can also skip encoding if we're just doing stream copy / transmuxing , although lpms_transcode doesn't support that right now (and when it does, we might want to do it in a separate code path for clarity). Once stream copy is in, we can also subsume lpms_rtmp2hls into the lpms_transcode and save another ~100 lines or so.

ericxtang commented:
Thanks for the answers @j0sh.  I'm not sure what you mean by "just take an existing filename" - but I remember chatting about removing the temporary file approach with you and @dob but we decided it's still useful to have the temporary files.  Maybe we can use some kind of naming convention to organize the video files so we can keep them around in case the user wants to save the video.  What do you think?

j0sh commented:
About the temporary files -- we'd still have them, but those would be written to disk around here https://github.com/livepeer/go-livepeer/blob/b54cc54975782608266ad44fdbbe509df261c58e/core/livepeernode.go#L240

rather than within the LPMS transcode function.

The LPMS transcoder signature could change from https://github.com/livepeer/lpms/blob/master/transcoder/ffmpeg_segment_transcoder.go#L27 to something like:

`func (t *FFMpegSegmentTranscoder) Transcode(fname string) ([][]byte, error)`

We can keep the return value as bytes for now, since I think changing that might require more thought.

The net effect would be to prevent writing additional test segments to disk, but it's a very minor issue since we don't yet have that many tests around transcoding.

j0sh commented:
Changed the API here https://github.com/livepeer/lpms/pull/52/commits/61305278a1d19e3028e02fbded6fb72c283b6cf9 and added tests here https://github.com/livepeer/lpms/pull/52/commits/444fb87005db5e9965a1f5da317d6868c0fb3e59

ericxtang commented:
:shipit: :shipit: :shipit:
